pkgs/test_api/lib/src/scaffolding/utils.dart

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import '../backend/invoker.dart';
+
+/// Returns a [Future] that completes after the [event loop][] has run the given
+/// number of [times] (20 by default).
+///
+/// [event loop]: https://medium.com/dartlang/dart-asynchronous-programming-isolates-and-event-loops-bffc3e296a6a
+///
+/// Awaiting this approximates waiting until all asynchronous work (other than
+/// work that's waiting for external resources) completes.
+Future pumpEventQueue({int times = 20}) {
+  if (times == 0) return Future.value();
+  // Use the event loop to allow the microtask queue to finish.
+  return Future(() => pumpEventQueue(times: times - 1));
+}
+
+/// Registers an exception that was caught for the current test.
+void registerException(Object error,
```

Comment:
cc @jakemac53 @grouma 

I didn't even know this existed until I stumbled across it while working on this change.

I've already excluded some deprecated APIs from `expect.dart` that are otherwise exposed through `test.dart`: https://github.com/dart-lang/test/blob/c5272043c5c6c524c3b3aaf5fa941e1dbb749b8d/pkgs/test_core/lib/test_core.dart#L12-L14

This way the new libraries come with extra API cleanup as we migrate.

We could do the same with `registerException` and avoid putting this in `scaffolding.dart` while retaining it in `test.dart` for backwards compatibility.

I like that both `fail` and `registerException` exist. They behave very differently - `fail` throws a specific type of exception to result in a test "failure", but not by communicating with the test framework directly - the exception needs to be able to bubble up the entire way. `registerException` avoids problems with `try/catch` in the user code because it reports to an error handler directly. It may result in a test "error", or a test "failure" if the exception type is specifically `TestFailure`. I don't think the distinction between "error" and "failure" is actually useful, and our users in practice are accustomed to being able to throw any exception to fail a test. https://github.com/dart-lang/test/issues/1465

Both `fail` and `registerException` have issues with error handling zones. This is a problem we could solve with `registerException`, but can't with `fail`.

Should we consider starting with a new API in `scaffolding.dart` to avoid these issues? For instance we can fairly easily work around the problem of error zones in user code by reporting to the test runner directly instead of reporting to a zone uncaught error handler.

Reply:
> Should we consider starting with a new API in scaffolding.dart to avoid these issues?

I think it makes sense. To be clear, this new API would not have the current "error" / "failure" distinction. Or is it required to support the existing `expect` use case?

Reply:
> To be clear, this new API would not have the current "error" / "failure" distinction.

Correct. I'm thinking that we'd drop `registerException` from `scaffolding.dart` and the user facing "API" is effectively "throw any exception". We'd hopefully drop the distinction between "error" and "failure" entirely.

Then in `hooks.dart` add `TestHandle.current.reportFailure(Object, [StackTrace?])` to take the place of `registerException`. This would take the place of `registerException` for the few edges where it is useful, and to report a failure past an error zone boundary.

The downside is there is no user visible API for reporting a failure past an error zone boundary since `hooks.dart` is not really considered user visible.

Reply:
For now I'm going to leave `registerException` out of both `scaffolding.dart` and `hooks.dart` and I won't worry yet to design a replacement.

The "official" way to report an error will be to throw an exception, and we'll look at ways that can also sneak through error zones in a followup down the road.